Mathlib/Combinatorics/SimpleGraph/FiveWheelLike.lean

```diff
@@ -0,0 +1,222 @@
+
+/--
+An `IsFiveWheelLike r k v w₁ w₂ s t` structure in `G` consists of vertices `v w₁ w₂` and `r`-sets
+`s` and `t` such that `{v, w₁, w₂}` induces the single edge `w₁w₂` (i.e. they form an
+`IsPathGraph3Compl`), `v, w₁, w₂ ∉ s ∪ t`, `s ∪ {v}, t ∪ {v}, s ∪ {w₁}, t ∪ {w₂}` are all
+`(r + 1)`- cliques and `#(s ∩ t) = k`. (If `G` is maximally `(r + 2)`-cliquefree and not complete
+ multipartite then `G` will contain such a structure : see
+`exists_isFiveWheelLike_of_maximal_cliqueFree_not_isCompleteMultipartite`.)
+-/
+structure IsFiveWheelLike (G : SimpleGraph α) (r k : ℕ) (v w₁ w₂ : α) (s t : Finset α) :
+    Prop where
+  /-- `{v, w₁, w₂}` induces the single edge `w₁w₂` -/
+  isPathGraph3Compl : G.IsPathGraph3Compl v w₁ w₂
+  not_mem_left : v ∉ s
+  not_mem_right : v ∉ t
+  fst_not_mem : w₁ ∉ s
+  snd_not_mem : w₂ ∉ t
+  isNClique_left : G.IsNClique (r + 1) (insert v s)
+  isNClique_fst_left : G.IsNClique (r + 1) (insert w₁ s)
+  isNClique_right : G.IsNClique (r + 1) (insert v t)
+  isNClique_snd_right : G.IsNClique (r + 1) (insert w₂ t)
+  card_eq : #(s ∩ t) = k
+
+lemma exists_isFiveWheelLike_of_maximal_cliqueFree_not_isCompleteMultipartite
+    (h : Maximal (fun H => H.CliqueFree (r + 2)) G) (hnc : ¬ G.IsCompleteMultipartite) :
+    ∃ v w₁ w₂ s t, G.IsFiveWheelLike r #(s ∩ t) v w₁ w₂ s t := by
+  obtain ⟨v, w₁, w₂, p3⟩ := exists_isPathGraph3Compl_of_not_isCompleteMultipartite hnc
+  obtain ⟨s, h1, h2, h3, h4⟩ := exists_of_maximal_cliqueFree_not_adj h p3.ne_fst p3.not_adj_fst
+  obtain ⟨t, h5, h6, h7, h8⟩ := exists_of_maximal_cliqueFree_not_adj h p3.ne_snd p3.not_adj_snd
+  exact  ⟨_, _, _, _, _, p3, h1, h5, h2, h6, h3, h4, h7, h8, rfl⟩
+
+/-- `G.FiveWheelLikeFree r k` means there is no `IsFiveWheelLike r k` structure in `G`. -/
+def FiveWheelLikeFree (G : SimpleGraph α) (r k : ℕ) : Prop :=
+    ∀ {v w₁ w₂ s t}, ¬ G.IsFiveWheelLike r k v w₁ w₂ s t
+
+namespace IsFiveWheelLike
+
+variable {v w₁ w₂ : α} {t : Finset α} (hw : G.IsFiveWheelLike r k v w₁ w₂ s t)
+
+include hw
+
+@[symm] lemma symm : G.IsFiveWheelLike r k v w₂ w₁ t s :=
+  let ⟨p2, d1, d2, d3, d4, c1, c2, c3, c4 , hk⟩ := hw
+  ⟨p2.symm, d2, d1, d4, d3, c3, c4, c1, c2, by rwa [inter_comm]⟩
+
+lemma fst_not_mem_right : w₁ ∉ t :=
+  fun h ↦ hw.isPathGraph3Compl.not_adj_fst <| hw.isNClique_right.1 (mem_insert_self ..)
+    (mem_insert_of_mem h) hw.isPathGraph3Compl.ne_fst
+
+lemma snd_not_mem_left : w₂ ∉ s := hw.symm.fst_not_mem_right
+
+/--
+Any graph containing an `IsFiveWheelLike r k` structure is not `(r + 1)`-colorable.
+-/
+lemma not_colorable_succ : ¬ G.Colorable (r + 1) := by
+  intro ⟨C⟩
+  have h := C.surjOn_of_card_le_isClique hw.isNClique_fst_left.1 (by simp [hw.isNClique_fst_left.2])
+  have := C.surjOn_of_card_le_isClique hw.isNClique_snd_right.1 (by simp [hw.isNClique_snd_right.2])
+-- Since `C` is an `r + 1`-coloring and `insert w₁ s` is an `r + 1`-clique, it contains a vertex `x`
+-- which shares its colour with `v`
```

Comment:
```suggestion
  -- Since `C` is an `r + 1`-coloring and `insert w₁ s` is an `r + 1`-clique, it contains a vertex `x`
  -- which shares its colour with `v`
```
and similarly below